deploy/deploy.go

```diff
@@ -146,7 +162,56 @@ type kubeVersion struct {
 	ServerVersion    *kversion.Info `json:"serverVersion,omitempty" yaml:"serverVersion,omitempty"`
 }
 
+// event turns the deployment into a cluster event protobuf.
+func (d *Deployment) event() *epb.Cluster {
+	c := &epb.Cluster{}
+	switch d.Cluster.(type) {
+	case *ExternalSpec:
+		c.Cluster = epb.Cluster_CLUSTER_TYPE_EXTERNAL
+	case *KindSpec:
+		c.Cluster = epb.Cluster_CLUSTER_TYPE_KIND
+	}
+	switch d.Ingress.(type) {
+	case *MetalLBSpec:
+		c.Ingress = epb.Cluster_INGRESS_TYPE_METALLB
+	}
+	switch d.CNI.(type) {
+	case *MeshnetSpec:
+		c.Cni = epb.Cluster_CNI_TYPE_MESHNET
+	}
+	for _, cntrl := range d.Controllers {
+		switch cntrl.(type) {
+		case *CEOSLabSpec:
+			c.Controllers = append(c.Controllers, epb.Cluster_CONTROLLER_TYPE_CEOSLAB)
+		case *IxiaTGSpec:
+			c.Controllers = append(c.Controllers, epb.Cluster_CONTROLLER_TYPE_IXIATG)
+		case *SRLinuxSpec:
+			c.Controllers = append(c.Controllers, epb.Cluster_CONTROLLER_TYPE_SRLINUX)
+		case *LemmingSpec:
+			c.Controllers = append(c.Controllers, epb.Cluster_CONTROLLER_TYPE_LEMMING)
+		}
+	}
+	return c
+}
+
 func (d *Deployment) Deploy(ctx context.Context, kubecfg string) (rerr error) {
+	if d.ReportUsage {
+		r, err := metrics.NewReporter(ctx, d.ReportUsageProjectID, d.ReportUsageTopicID)
```

Comment:
Put the entire thing in a method so you can say:

```
// reportUsage returns a function to defer by the caller.
//
//      if d.ReportUsage {
//              finish = d.reportUsage(ctx)
//              defer finish(rerr)
//      }
func (d *Deployment) reportUsage(ctx context.Context) func(error) {
        r, err := metrics.NewReporter(ctx, d.ReportUsageProjectID, d.ReportUsageTopicID)
        if err != nil {
                log.Warningf("Unable to create metrics reporter: %v", err)
                return func(_ error) {}
        }
        if id, err := r.ReportDeployClusterStart(ctx, d.event()); err != nil {
                log.Warningf("Unable to report cluster deployment start event: %v", err)
                return func(_ error) { r.Close() }
        }
        return func(rerr error) {
                if err := r.ReportDeployClusterEnd(ctx, id, rerr); err != nil {
                        log.Warningf("Unable to report cluster deployment end event: %v", err)
                }
        }
}
```

Reply:
done